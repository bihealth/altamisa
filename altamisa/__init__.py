# -*- coding: utf-8 -*-
"""Top-level package for AltamISA."""

__author__ = 'Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>'
# TODO: add version with versioneer
