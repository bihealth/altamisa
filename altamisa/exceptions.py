# -*- coding: utf-8 -*-
"""
Exceptions and Warnings used in the AltamISA library.
"""

__author__ = "Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>"


class IsaException(Exception):
    """Base class for exceptions raised by Altamisa."""


class ParseIsatabException(IsaException):
    """Exception raised on problems parsing ISA-TAB."""


class WriteIsatabException(IsaException):
    """Exception raised on problems writing ISA-TAB."""


class IsaWarning(Warning):
    """Base class for warnings raised by Altamisa."""


class ParseIsatabWarning(IsaWarning):
    """Warning raised on problems parsing ISA-TAB."""


class WriteIsatabWarning(IsaWarning):
    """Warning raised on problems writing ISA-TAB."""


class IsaValidationWarning(IsaWarning):
    """Warning raised on problems validating ISA models or objects."""


class AdvisoryIsaValidationWarning(IsaValidationWarning):
    """Warning raised on uncritical problems when validating ISA models or objects."""


class ModerateIsaValidationWarning(IsaValidationWarning):
    """Warning raised on moderate problems when validating ISA models or objects."""


class CriticalIsaValidationWarning(IsaValidationWarning):
    """Warning raised on critical problems when validating ISA models or objects."""
