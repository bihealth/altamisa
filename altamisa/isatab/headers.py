# -*- coding: utf-8 -*-
"""This module contains code for the representation of headers from study and
assay files and parsing thereof.
"""

from __future__ import generator_stop
from typing import Iterator, List

from ..constants import table_headers
from ..exceptions import ParseIsatabException


__author__ = "Manuel Holtgrewe <manuel.holtgrewe@bihealth.de>"


class ColumnHeader:
    """Column header in a study or assay file"""

    def __init__(self, column_type, col_no, span):
        #: The type of this header
        self.column_type = column_type
        #: The column number this header refers to
        self.col_no = col_no
        #: Number of columns this header spans
        self.span = span
        #: Link to the TermSourceRefHeader to use
        self.term_source_ref_header = None
        #: Link to the UnitHeader to use
        self.unit_header = None

    def __str__(self):
        tpl = "ColumnHeader(column_type={}, col_no={}, span={})"
        return tpl.format(*map(repr, (self.column_type, self.col_no, self.span)))

    def __repr__(self):
        return str(self)

    def get_simple_string(self) -> List[str]:
        """Return a list of simple string representations of the column types"""
        return [self.column_type]


class SimpleColumnHeader(ColumnHeader):
    """Base class for simple column headers."""

    #: The value to use for the ``type`` argument.
    column_type = None

    def __init__(self, col_no):
        super().__init__(self.column_type, col_no, 1)


# Material header


class ExtractHeader(SimpleColumnHeader):
    """Extract header in an assay"""

    column_type = table_headers.EXTRACT_NAME


class LabeledExtractHeader(SimpleColumnHeader):
    """Labeled Extract header in an assay"""

    column_type = table_headers.LABELED_EXTRACT_NAME


class LibraryHeader(SimpleColumnHeader):
    """Library header in an assay"""

    column_type = table_headers.LIBRARY_NAME


class SampleHeader(SimpleColumnHeader):
    """Sample header in a study or assay"""

    column_type = table_headers.SAMPLE_NAME


class SourceHeader(SimpleColumnHeader):
    """Source header in a study"""

    column_type = table_headers.SOURCE_NAME


# Data header


class ArrayDataFileHeader(SimpleColumnHeader):
    """ArrayData header in an assay"""

    column_type = table_headers.ARRAY_DATA_FILE


class ArrayDataMatrixFileHeader(SimpleColumnHeader):
    """ArrayData Matrix File header in an assay"""

    column_type = table_headers.ARRAY_DATA_MATRIX_FILE


class ArrayDesignFileHeader(SimpleColumnHeader):
    """ArrayDesignFile header in an assay"""

    column_type = table_headers.ARRAY_DESIGN_FILE


class DerivedArrayDataFileHeader(SimpleColumnHeader):
    """DerivedArrayData header in an assay"""

    column_type = table_headers.DERIVED_ARRAY_DATA_FILE


class DerivedArrayDataMatrixFileHeader(SimpleColumnHeader):
    """DerivedArrayData header in an assay"""

    column_type = table_headers.DERIVED_ARRAY_DATA_MATRIX_FILE


class DerivedDataFileHeader(SimpleColumnHeader):
    """Derived Data File header in an assay"""

    column_type = table_headers.DERIVED_DATA_FILE


class DerivedSpectralDataFileHeader(SimpleColumnHeader):
    """DerivedSpectralData header in an assay"""

    column_type = table_headers.DERIVED_SPECTRAL_DATA_FILE


class ImageFileHeader(SimpleColumnHeader):
    """Image File header in an assay"""

    column_type = table_headers.IMAGE_FILE


class MetaboliteAssignmentFileHeader(SimpleColumnHeader):
    """PeptideAssignment header in an assay"""

    column_type = table_headers.METABOLITE_ASSIGNMENT_FILE


class PeptideAssignmentFileHeader(SimpleColumnHeader):
    """PeptideAssignment header in an assay"""

    column_type = table_headers.PEPTIDE_ASSIGNMENT_FILE


class PostTranslationalModificationAssignmentFileHeader(SimpleColumnHeader):
    """PostTranslationalModificationAssignment header in an assay"""

    column_type = table_headers.POST_TRANSLATIONAL_MODIFICATION_ASSIGNMENT_FILE


class ProteinAssignmentFileHeader(SimpleColumnHeader):
    """ProteinAssignment header in an assay"""

    column_type = table_headers.PROTEIN_ASSIGNMENT_FILE


class RawDataFileHeader(SimpleColumnHeader):
    """Raw Data header in an assay"""

    column_type = table_headers.RAW_DATA_FILE


class RawSpectralDataFileHeader(SimpleColumnHeader):
    """Raw Spectral Data header in an assay"""

    column_type = table_headers.RAW_SPECTRAL_DATA_FILE


class SpotPickingFileHeader(SimpleColumnHeader):
    """SpotPickingFile header in an assay"""

    column_type = table_headers.SPOT_PICKING_FILE


# Assay header


class AssayNameHeader(SimpleColumnHeader):
    """Assay Name header in an assay"""

    column_type = table_headers.ASSAY_NAME


class DataTransformationNameHeader(SimpleColumnHeader):
    """DataTransformationName header in an assay"""

    column_type = table_headers.DATA_TRANSFORMATION_NAME


class GelElectrophoresisAssayNameHeader(SimpleColumnHeader):
    """GelElectrophoresisAssayName header in an assay"""

    column_type = table_headers.GEL_ELECTROPHORESIS_ASSAY_NAME


class HybridizationAssayNameHeader(SimpleColumnHeader):
    """HybridizationAssayName header in an assay"""

    column_type = table_headers.HYBRIDIZATION_ASSAY_NAME


class MsAssayNameHeader(SimpleColumnHeader):
    """MsAssayName header in an assay"""

    column_type = table_headers.MS_ASSAY_NAME


class NormalizationNameHeader(SimpleColumnHeader):
    """Normalization Name header in a assay"""

    column_type = table_headers.NORMALIZATION_NAME


class ProtocolRefHeader(SimpleColumnHeader):
    """Protocol REF header in a study or assay"""

    column_type = table_headers.PROTOCOL_REF


class ScanNameHeader(SimpleColumnHeader):
    """ScanName header in assay"""

    column_type = table_headers.SCAN_NAME


# Other simple header


class ArrayDesignRefHeader(SimpleColumnHeader):
    """ArrayDesignRef header in an assay"""

    column_type = table_headers.ARRAY_DESIGN_REF


class DateHeader(SimpleColumnHeader):
    """Date annotation header in a study or assay"""

    column_type = table_headers.DATE


class FirstDimensionHeader(SimpleColumnHeader):
    """First Dimension header in an assay"""

    column_type = table_headers.FIRST_DIMENSION


class LabelHeader(SimpleColumnHeader):
    """Label header in an assay"""

    column_type = table_headers.LABEL


class MaterialTypeHeader(SimpleColumnHeader):
    """Material Type header in an assay"""

    column_type = table_headers.MATERIAL_TYPE


class PerformerHeader(SimpleColumnHeader):
    """Performer header in an assay"""

    column_type = table_headers.PERFORMER


class SecondDimensionHeader(SimpleColumnHeader):
    """Second Dimension header in an assay"""

    column_type = table_headers.SECOND_DIMENSION


class TermRefAnnotationHeader(ColumnHeader):
    """Term reference annotation header"""

    def __init__(self, col_no):
        super().__init__(table_headers.TERM_SOURCE_REF, col_no, 2)

    def get_simple_string(self) -> List[str]:
        """Return a list of simple string representations of the column types"""
        return [table_headers.TERM_SOURCE_REF, table_headers.TERM_ACCESSION_NUMBER]


class UnitHeader(SimpleColumnHeader):
    """Unit annotation header in a study or assay"""

    column_type = table_headers.UNIT


# Labeled header


class LabeledColumnHeader(ColumnHeader):
    """Base class for labeled column headers."""

    #: The value to use for the ``type`` argument.
    column_type = None

    def __init__(self, col_no, label):
        super().__init__(self.column_type, col_no, 1)
        self.label = label

    def __str__(self):
        tpl = "LabeledColumnHeader(column_type={}, col_no={}, span={}, label={})"
        return tpl.format(*map(repr, (self.column_type, self.col_no, self.span, self.label)))

    def __repr__(self):
        return str(self)

    def get_simple_string(self):
        """Return a list of simple string representations of the column types"""
        return ["".join((self.column_type, "[", self.label, "]"))]


class CharacteristicsHeader(LabeledColumnHeader):
    """Material ``Characteristics[*]`` header in a study or assay"""

    column_type = table_headers.CHARACTERISTICS


class CommentHeader(LabeledColumnHeader):
    """``Comment`` header in a study or assay"""

    column_type = table_headers.COMMENT


class FactorValueHeader(LabeledColumnHeader):
    """``Factor Value[*]`` header in a study or assay"""

    column_type = table_headers.FACTOR_VALUE


class ParameterValueHeader(LabeledColumnHeader):
    """Protocol ``Parameter Value[*]`` header in a study or assay"""

    column_type = table_headers.PARAMETER_VALUE


# Header parsers


class HeaderParserBase:
    """
    Helper base class for parsing a header from a study or assay file.

    :type tokens: list
    :param tokens: List of strings, e.g. a split line read from a tsv/cvs file.
    """

    #: Names of the allowed headers
    allowed_headers = None

    #: Headers that are mapped to ``SimpleColumnHeader``
    simple_headers = {
        # Material headers
        table_headers.EXTRACT_NAME: ExtractHeader,
        table_headers.LABELED_EXTRACT_NAME: LabeledExtractHeader,
        table_headers.LIBRARY_NAME: LibraryHeader,
        table_headers.SAMPLE_NAME: SampleHeader,
        table_headers.SOURCE_NAME: SourceHeader,
        # Data headers
        table_headers.ARRAY_DATA_FILE: ArrayDataFileHeader,
        table_headers.ARRAY_DATA_MATRIX_FILE: ArrayDataMatrixFileHeader,
        table_headers.ARRAY_DESIGN_FILE: ArrayDesignFileHeader,
        table_headers.DERIVED_ARRAY_DATA_FILE: DerivedArrayDataFileHeader,
        table_headers.DERIVED_ARRAY_DATA_MATRIX_FILE: DerivedArrayDataMatrixFileHeader,
        table_headers.DERIVED_DATA_FILE: DerivedDataFileHeader,
        table_headers.DERIVED_SPECTRAL_DATA_FILE: DerivedSpectralDataFileHeader,
        table_headers.IMAGE_FILE: ImageFileHeader,
        table_headers.METABOLITE_ASSIGNMENT_FILE: MetaboliteAssignmentFileHeader,
        table_headers.PEPTIDE_ASSIGNMENT_FILE: PeptideAssignmentFileHeader,
        table_headers.POST_TRANSLATIONAL_MODIFICATION_ASSIGNMENT_FILE: PostTranslationalModificationAssignmentFileHeader,
        table_headers.PROTEIN_ASSIGNMENT_FILE: ProteinAssignmentFileHeader,
        table_headers.RAW_DATA_FILE: RawDataFileHeader,
        table_headers.RAW_SPECTRAL_DATA_FILE: RawSpectralDataFileHeader,
        table_headers.SPOT_PICKING_FILE: SpotPickingFileHeader,
        # Process names or Ref
        table_headers.ASSAY_NAME: AssayNameHeader,
        table_headers.DATA_TRANSFORMATION_NAME: DataTransformationNameHeader,
        table_headers.GEL_ELECTROPHORESIS_ASSAY_NAME: GelElectrophoresisAssayNameHeader,
        table_headers.HYBRIDIZATION_ASSAY_NAME: HybridizationAssayNameHeader,
        table_headers.MS_ASSAY_NAME: MsAssayNameHeader,
        table_headers.NORMALIZATION_NAME: NormalizationNameHeader,
        table_headers.PROTOCOL_REF: ProtocolRefHeader,
        table_headers.SCAN_NAME: ScanNameHeader,
        # Simple headers
        table_headers.ARRAY_DESIGN_REF: ArrayDesignRefHeader,
        table_headers.DATE: DateHeader,
        table_headers.FIRST_DIMENSION: FirstDimensionHeader,
        table_headers.LABEL: LabelHeader,
        table_headers.MATERIAL_TYPE: MaterialTypeHeader,
        table_headers.PERFORMER: PerformerHeader,
        table_headers.SECOND_DIMENSION: SecondDimensionHeader,
        # Secondary annotations
        table_headers.UNIT: UnitHeader,
    }

    #: Labeled headers
    labeled_headers = {
        table_headers.CHARACTERISTICS: CharacteristicsHeader,
        table_headers.COMMENT: CommentHeader,
        table_headers.FACTOR_VALUE: FactorValueHeader,
        table_headers.PARAMETER_VALUE: ParameterValueHeader,
    }

    def __init__(self, tokens):
        self.tokens = tokens
        self.it = iter(tokens)
        self.col_no = 0

    def run(self) -> Iterator[ColumnHeader]:
        """Parse the header"""
        while True:
            try:
                yield self._parse_next()
            except StopIteration:
                break

    def _parse_next(self):
        # Get next value from header
        val = next(self.it)  # StopIteration is OK here
        # Process either by exact match to "Term Source REF", or other exact
        # matches, or any of the prefix matches (e.g., "Comment[Label])"
        if val == table_headers.TERM_SOURCE_REF:
            return self._parse_term_source_ref()
        elif val in self.simple_headers:
            if val not in self.allowed_headers:
                tpl = 'Header "{}" not allowed in {}.'
                msg = tpl.format(val, self.file_type)
                raise ParseIsatabException(msg)
            return self._parse_simple_column_header(self.simple_headers[val])
        else:
            for label, type_ in self.labeled_headers.items():
                if val.startswith(label):
                    if label not in self.allowed_headers:
                        tpl = 'Header "{}" not allowed in {}.'
                        msg = tpl.format(label, self.file_type)
                        raise ParseIsatabException(msg)
                    return self._parse_labeled_column_header(val, label, type_)
        # None of the if-statements above was taken
        tpl = 'Header "{}" unknown, processing unclear'
        msg = tpl.format(val)
        raise ParseIsatabException(msg)

    def _parse_term_source_ref(self):
        # Getting a StopIteration here is NOT okay, there must be a column
        # after "Term Source REF" giving the ontology the term is from.
        try:
            val = next(self.it)
        except StopIteration as e:
            msg = 'Expected one more column on seeing "Term Source REF"'
            raise ParseIsatabException(msg) from e
        if val != table_headers.TERM_ACCESSION_NUMBER:
            msg = 'Expected column "Term Accession Number" after seeing "Term Source REF"'
            raise ParseIsatabException(msg)
        self.col_no += 2
        return TermRefAnnotationHeader(self.col_no - 2)

    def _parse_simple_column_header(self, type_):
        self.col_no += 1
        return type_(self.col_no - 1)

    def _parse_labeled_column_header(self, val, key, type_):
        tok = val[len(key) :]  # strip '^{key}'
        if not tok or tok[0] != "[" or tok[-1] != "]":
            tpl = "Problem parsing labeled header {}"
            msg = tpl.format(val)
            raise ParseIsatabException(msg)
        self.col_no += 1
        return type_(self.col_no - 1, tok[1:-1])


class StudyHeaderParser(HeaderParserBase):
    """Helper class for parsing header of a study or assay."""

    file_type = "study"  # for exceptions only

    allowed_headers = (
        # Material names
        table_headers.SAMPLE_NAME,
        table_headers.SOURCE_NAME,
        # Process names
        table_headers.PROTOCOL_REF,
        # Simple headers
        table_headers.DATE,
        table_headers.PERFORMER,
        # Labeled headers
        table_headers.CHARACTERISTICS,
        table_headers.COMMENT,
        table_headers.FACTOR_VALUE,
        table_headers.PARAMETER_VALUE,
        # Secondary annotations
        table_headers.TERM_SOURCE_REF,
        table_headers.UNIT,
    )


class AssayHeaderParser(HeaderParserBase):
    """Helper class for parsing header of a assay file."""

    file_type = "assay"  # for exceptions only

    allowed_headers = (
        # Material names
        table_headers.EXTRACT_NAME,
        table_headers.LABELED_EXTRACT_NAME,
        table_headers.LIBRARY_NAME,
        table_headers.SAMPLE_NAME,
        # Data names
        table_headers.ARRAY_DATA_FILE,
        table_headers.ARRAY_DATA_MATRIX_FILE,
        table_headers.ARRAY_DESIGN_FILE,
        table_headers.DERIVED_ARRAY_DATA_FILE,
        table_headers.DERIVED_ARRAY_DATA_MATRIX_FILE,
        table_headers.DERIVED_DATA_FILE,
        table_headers.DERIVED_SPECTRAL_DATA_FILE,
        table_headers.IMAGE_FILE,
        table_headers.METABOLITE_ASSIGNMENT_FILE,
        table_headers.PEPTIDE_ASSIGNMENT_FILE,
        table_headers.POST_TRANSLATIONAL_MODIFICATION_ASSIGNMENT_FILE,
        table_headers.PROTEIN_ASSIGNMENT_FILE,
        table_headers.RAW_DATA_FILE,
        table_headers.RAW_SPECTRAL_DATA_FILE,
        table_headers.SPOT_PICKING_FILE,
        # Process names or Ref
        table_headers.ASSAY_NAME,
        table_headers.DATA_TRANSFORMATION_NAME,
        table_headers.GEL_ELECTROPHORESIS_ASSAY_NAME,
        table_headers.HYBRIDIZATION_ASSAY_NAME,
        table_headers.MS_ASSAY_NAME,
        table_headers.NORMALIZATION_NAME,
        table_headers.PROTOCOL_REF,
        table_headers.SCAN_NAME,
        # Simple headers
        table_headers.ARRAY_DESIGN_REF,
        table_headers.DATE,
        table_headers.FIRST_DIMENSION,
        table_headers.LABEL,
        table_headers.MATERIAL_TYPE,
        table_headers.PERFORMER,
        table_headers.SECOND_DIMENSION,
        # Labeled headers
        table_headers.CHARACTERISTICS,
        table_headers.COMMENT,
        table_headers.PARAMETER_VALUE,
        # Secondary annotations
        table_headers.TERM_SOURCE_REF,
        table_headers.UNIT,
    )
