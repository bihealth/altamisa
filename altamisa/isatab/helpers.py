# -*- coding: utf-8 -*-
"""
Module for helper functions of global use.
"""


import warnings


from ..exceptions import ParseIsatabWarning


__author__ = "Mathias Kuhring <mathias.kuhring@bihealth.de>"


def is_ontology_term_ref(v):
    """Duck typing check for objects of class `models.OntologyTermRef`"""
    return hasattr(v, "name") and hasattr(v, "ontology_name") and hasattr(v, "accession")


def list_strip(line: list):
    """Remove trailing space from strings in a list (e.g. a csv line)"""
    new_line = [field.strip() for field in line]
    if new_line != line:
        tpl = "Removed trailing whitespaces in fields of line: {}"
        msg = tpl.format(line)
        warnings.warn(msg, ParseIsatabWarning)
    return new_line
