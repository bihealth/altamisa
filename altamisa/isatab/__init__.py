# -*- coding: utf-8 -*-
"""Module for parsing ISA-Tab files."""

# Make all models and the ``*Reader`` classes visible within this module.

from .headers import *  # noqa: F403, F401
from .models import *  # noqa: F403, F401
from .parse_assay_study import (  # noqa: F401
    AssayReader,
    AssayRowReader,
    StudyReader,
    StudyRowReader,
)
from .parse_investigation import InvestigationReader  # noqa: F401
from .validate_assay_study import AssayValidator, StudyValidator  # noqa: F401
from .validate_investigation import InvestigationValidator  # noqa: F401
from .write_assay_study import AssayWriter, StudyWriter, RefTableBuilder  # noqa: F401
from .write_investigation import InvestigationWriter  # noqa: F401
