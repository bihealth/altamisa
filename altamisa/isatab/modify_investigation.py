from pathlib import Path
from typing import Union

import attr

from altamisa.isatab import InvestigationReader

__author__ = (
    "Thomas Sell <thomas.sell@charite.de>"
)

class InvestigationForge:
    """
    Provides methods to add assays to an existing investigation.

    :type input_path: str | Path
    :param input_path: Location of investigation to modify.
    """

    def __init__(self, input_path: Union[str, Path]):
        i_file = Path(input_path).expanduser().resolve()

        with i_file.open("rt") as f:
            self.investigation = InvestigationReader.from_stream(f).read()

        if len(self.investigation.studies) != 1:
            # TODO: add support for multiple studies
            raise IndexError("Only single study investigations are supported.")

    @staticmethod
    def _join_protocols(x: dict, y: dict) -> dict:
        """
        Join two dicts of ISA protocols.
        For duplicate protocols, join parameter dicts.
        """
        for protocol in y:
            if protocol not in x:
                # no conflict > add protocol
                x.update({protocol: y[protocol]})
            else:
                # same protocol > check parameters
                old_parameters = x[protocol].parameters
                new_parameters = y[protocol].parameters
                for parameter in new_parameters:
                    if parameter not in old_parameters:
                        # no conflict > add parameter
                        old_parameters.update({parameter: new_parameters[parameter]})
                    else:
                        # existing parameter: raise for ontology mismatch
                        if new_parameters[parameter] != old_parameters[parameter]:
                            raise ValueError(
                                f'Ontology term mismatch for parameter "{parameter}" of protocol "{protocol}".'
                            )
        return x

    def add_assay(self, input_path: Union[str, Path]):
        """
        Add assay to investigation file.

        :type input_path: str | Path
        :param input_path: Location of bare assay definition to add.
        :rtype: models.InvestigationInfo
        :returns: Investigation model including all information from the investigation file.
        """
        new_assay = Path(input_path).expanduser().resolve()
        with new_assay.open("rt") as f:
            reader = InvestigationReader(f)
            new_assays = tuple(a for a in reader._read_study_assays())
            new_protocols = {p.name: p for p in reader._read_study_protocols()}

        old_protocols = self.investigation.studies[0].protocols
        joined_protocols = self._join_protocols(
            old_protocols,
            new_protocols,
        )

        assays = self.investigation.studies[0].assays
        assays += new_assays

        updated_study = attr.evolve(self.investigation.studies[0], protocols=joined_protocols, assays=assays)
        self.investigation = attr.evolve(self.investigation, studies=(updated_study,))

    def add_protocol(self):
        raise NotImplementedError

    def remove_protocol(self):
        raise NotImplementedError
