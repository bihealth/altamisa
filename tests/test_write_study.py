# -*- coding: utf-8 -*-
"""Tests for writing ISA study files"""


import filecmp
import pytest
import os

from altamisa.exceptions import ModerateIsaValidationWarning, IsaWarning, ParseIsatabWarning
from altamisa.isatab import (
    InvestigationReader,
    InvestigationValidator,
    StudyReader,
    StudyValidator,
    StudyWriter,
)


# Helper to load, write and compare studies
def _parse_write_assert(investigation_file, tmp_path, quote=None):
    # Load investigation
    investigation = InvestigationReader.from_stream(investigation_file).read()
    InvestigationValidator(investigation).validate()
    directory = os.path.normpath(os.path.dirname(investigation_file.name))
    # Iterate studies
    for s, study_info in enumerate(investigation.studies):
        # Load study
        path_in = os.path.join(directory, study_info.info.path)
        with open(path_in, "rt") as inputf:
            study = StudyReader.from_stream("S{}".format(s + 1), inputf).read()
        StudyValidator(investigation, study_info, study).validate()
        # Write study to temporary file
        path_out = tmp_path / study_info.info.path
        with open(path_out, "wt", newline="") as file:
            StudyWriter.from_stream(study, file, quote=quote).write()
        assert filecmp.cmp(path_in, path_out, shallow=False)


def test_study_writer_minimal(minimal_investigation_file, tmp_path):
    with pytest.warns(IsaWarning) as record:
        _parse_write_assert(minimal_investigation_file, tmp_path)
    # Check warnings
    assert 2 == len(record)


def test_study_writer_minimal2(minimal2_investigation_file, tmp_path):
    with pytest.warns(IsaWarning) as record:
        _parse_write_assert(minimal2_investigation_file, tmp_path)
    # Check warnings
    assert 2 == len(record)


def test_study_writer_small(small_investigation_file, tmp_path):
    with pytest.warns(IsaWarning) as record:
        _parse_write_assert(small_investigation_file, tmp_path)
    # Check warnings
    assert 2 == len(record)


def test_study_writer_small2(small2_investigation_file, tmp_path):
    with pytest.warns(IsaWarning) as record:
        _parse_write_assert(small2_investigation_file, tmp_path)
    # Check warnings
    assert 1 == len(record)


def test_study_writer_BII_I_1(BII_I_1_investigation_file, tmp_path):
    with pytest.warns(IsaWarning) as record:
        _parse_write_assert(BII_I_1_investigation_file, tmp_path, quote='"')
    # Check warnings
    assert 1 == len(record)


def test_study_writer_gelelect(gelelect_investigation_file, tmp_path):
    with pytest.warns(IsaWarning) as record:
        _parse_write_assert(gelelect_investigation_file, tmp_path, quote='"')
    # Check warnings
    assert 3 == len(record)
    msg = "Skipping empty ontology source: , , , "
    assert record[0].category == ParseIsatabWarning
    assert str(record[0].message) == msg
    msg = (
        "Investigation with only one study contains metadata:\n\tID:\t1551099271112\n\tTitle:\t"
        "Investigation\n\tPath:\ti_Investigation.txt\n\tSubmission Date:\t\n\tPublic Release "
        "Date:\tNone\n\tPrefer recording metadata in the study section."
    )
    assert record[1].category == ModerateIsaValidationWarning
    assert str(record[1].message) == msg
    msg = "Study without title:\nID:\tstudy01\nTitle:\t\nPath:\ts_study01.txt"
    assert record[2].category == ModerateIsaValidationWarning
    assert str(record[2].message) == msg
