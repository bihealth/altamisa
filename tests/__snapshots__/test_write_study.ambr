# serializer version: 1
# name: test_study_writer_BII_I_1
  list([
    'Skipping empty ontology source: , , , ',
  ])
# ---
# name: test_study_writer_gelelect
  list([
    "<class 'altamisa.exceptions.ParseIsatabWarning'>",
    "<class 'altamisa.exceptions.ModerateIsaValidationWarning'>",
    "<class 'altamisa.exceptions.ModerateIsaValidationWarning'>",
  ])
# ---
# name: test_study_writer_gelelect.1
  list([
    'Skipping empty ontology source: , , , ',
    '''
      Investigation with only one study contains metadata:
      	ID:	1551099271112
      	Title:	Investigation
      	Path:	i_Investigation.txt
      	Description:	
      	Submission Date:	None
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
    '''
      Study without title:
      ID:	study01
      Title:	
      Path:	s_study01.txt
    ''',
  ])
# ---
# name: test_study_writer_minimal
  list([
    '''
      Investigation with only one study contains metadata:
      	ID:	i_minimal
      	Title:	Minimal Investigation
      	Path:	i_minimal.txt
      	Description:	
      	Submission Date:	None
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
    '''
      Assay without platform:
      Path:	a_minimal.txt
      Measurement Type:	exome sequencing assay
      Technology Type:	nucleotide sequencing
      Technology Platform:	
    ''',
  ])
# ---
# name: test_study_writer_minimal2
  list([
    '''
      Investigation with only one study contains metadata:
      	ID:	i_minimal
      	Title:	Minimal Investigation
      	Path:	i_minimal2.txt
      	Description:	
      	Submission Date:	None
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
    '''
      Assay without platform:
      Path:	a_minimal2.txt
      Measurement Type:	exome sequencing assay
      Technology Type:	nucleotide sequencing
      Technology Platform:	
    ''',
  ])
# ---
# name: test_study_writer_small
  list([
    '''
      Investigation with only one study contains metadata:
      	ID:	i_small
      	Title:	Small Investigation
      	Path:	i_small.txt
      	Description:	
      	Submission Date:	None
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
    '''
      Assay without platform:
      Path:	a_small.txt
      Measurement Type:	exome sequencing assay
      Technology Type:	nucleotide sequencing
      Technology Platform:	
    ''',
    "Found annotated material/file without name: Material(type='Sample Name', unique_name='S1-Empty Sample Name-13-5', name='', extract_label=None, characteristics=(Characteristics(name='status', value=['1'], unit=None),), comments=(), factor_values=(FactorValue(name='treatment', value=[''], unit=None),), material_type=None, headers=['Sample Name', 'Characteristics[status]', 'Factor Value[treatment]'])",
  ])
# ---
# name: test_study_writer_small2
  list([
    '''
      Investigation with only one study contains metadata:
      	ID:	i_small2
      	Title:	Small Investigation
      	Path:	i_small2.txt
      	Description:	
      	Submission Date:	None
      	Public Release Date:	None
      	Prefer recording metadata in the study section.
    ''',
  ])
# ---
